theseus/src/state/children.rs

```diff
@@ -1,37 +1,159 @@
+use futures::stream::{self, StreamExt};
 use std::{collections::HashMap, sync::Arc};
-use tokio::process::Child;
+use tokio::io::{AsyncBufReadExt, BufReader};
+use tokio::process::{ChildStderr, ChildStdout};
 use tokio::sync::RwLock;
 
 // Child processes (instances of Minecraft)
-// A wrapper over a Hashmap connecting PID -> Child
-// Left open for future functionality re: polling children
-pub struct Children(HashMap<u32, Arc<RwLock<Child>>>);
+// A wrapper over a Hashmap connecting PID -> MinecraftChild
+pub struct Children(HashMap<u32, Arc<RwLock<MinecraftChild>>>);
+
+// Minecraft Child, bundles together the PID, the actual Child, and the easily queryable stdout and stderr streams
+#[derive(Debug)]
+pub struct MinecraftChild {
+    pub pid: u32,
+    pub child: tokio::process::Child,
+    pub stdout: SharedOutput,
+    pub stderr: SharedOutput,
+}
 
 impl Children {
     pub fn new() -> Children {
         Children(HashMap::new())
     }
 
-    // Inserts and returns a ref to the child
-    // Unlike a Hashmap, this directly returns the reference to the Child rather than any previously stored Child that may exist
-    pub fn insert(
+    // Inserts a child process to keep track of, and returns a reference to the container struct MinecraftChild
+    // The threads for stdout and stderr are spawned here
+    // Unlike a Hashmap's 'insert', this directly returns the reference to the Child rather than any previously stored Child that may exist
+    pub fn insert_process(
         &mut self,
         pid: u32,
-        child: tokio::process::Child,
-    ) -> Arc<RwLock<Child>> {
-        let child = Arc::new(RwLock::new(child));
-        self.0.insert(pid, child.clone());
-        child
+        mut child: tokio::process::Child,
+    ) -> Arc<RwLock<MinecraftChild>> {
+        // Create std watcher threads for stdout and stderr
+        let stdout = SharedOutput::new();
+        if let Some(child_stdout) = child.stdout.take() {
+            let stdout_clone = stdout.clone();
+            tokio::spawn(async move {
+                stdout_clone.read_stdout(child_stdout).await.unwrap();
```

Comment:
no unwraps! properly handle errors here. by the way, does this sync up? like will this constantly stream input? the code seems to be just one time